probe-rs/src/probe/fake_probe.rs

```diff
@@ -80,90 +86,96 @@ impl SwdSequence for &mut MockCore {
-        todo!()
+    fn read_16(&mut self, address: u64, data: &mut [u16]) -> Result<(), ArmError> {
+        self.arm_read_handler.as_ref().unwrap()(address, unsafe {
+            std::slice::from_raw_parts_mut(data.as_mut_ptr() as *mut u8, data.len() * 2)
+        })
     }
 
     fn read_32(&mut self, address: u64, data: &mut [u32]) -> Result<(), ArmError> {
-        for (i, val) in data.iter_mut().enumerate() {
-            let address = address + (i as u64 * 4);
-
-            match address {
-                // DHCSR
-                Dhcsr::ADDRESS_OFFSET => {
-                    let mut dhcsr: u32 = self.dhcsr.into();
+        match address {
+            // DHCSR
+            Dhcsr::ADDRESS_OFFSET => {
+                let mut dhcsr: u32 = self.dhcsr.into();
 
-                    if self.is_halted {
-                        dhcsr |= 1 << 17;
-                    }
-
-                    // Always set S_REGRDY, and say that a register value can
-                    // be read.
-                    dhcsr |= 1 << 16;
-
-                    *val = dhcsr;
-                    println!("Read  DHCSR: {:#x} = {:#x}", address, val);
+                if self.is_halted {
+                    dhcsr |= 1 << 17;
                 }
 
-                _ => {
-                    *val = 0;
-                    println!("Read {:#010x} = 0", address);
-                }
+                // Always set S_REGRDY, and say that a register value can
+                // be read.
+                dhcsr |= 1 << 16;
+
+                data[0] = dhcsr;
+                tracing::trace!("[read_32] <DHCSR>: {:#x} = {:#x}", address, data[0]);
+                Ok(())
+            }
+            0xE0000000..=0xE00FFFFF => {
+                tracing::trace!("[read_32] <debug protocol>: {:08x} = {}", address, data[0]);
+                Ok(())
             }
+            _ => self.arm_read_handler.as_ref().unwrap()(address, unsafe {
+                std::slice::from_raw_parts_mut(data.as_mut_ptr() as *mut u8, data.len() * 4)
+            }),
         }
-
-        Ok(())
     }
 
-    fn read_64(&mut self, _address: u64, _data: &mut [u64]) -> Result<(), ArmError> {
-        todo!()
+    fn read_64(&mut self, address: u64, data: &mut [u64]) -> Result<(), ArmError> {
+        self.arm_read_handler.as_ref().unwrap()(address, unsafe {
+            std::slice::from_raw_parts_mut(data.as_mut_ptr() as *mut u8, data.len() * 8)
+        })
     }
 
-    fn write_8(&mut self, _address: u64, _data: &[u8]) -> Result<(), ArmError> {
-        todo!()
+    fn write_8(&mut self, address: u64, data: &[u8]) -> Result<(), ArmError> {
+        self.arm_write_handler.as_ref().unwrap()(address, data)
     }
 
-    fn write_16(&mut self, _address: u64, _data: &[u16]) -> Result<(), ArmError> {
-        todo!()
+    fn write_16(&mut self, address: u64, data: &[u16]) -> Result<(), ArmError> {
+        self.arm_write_handler.as_ref().unwrap()(address, unsafe {
```

Comment:
Please don't use unsafe. There is no reason for this.

Reply:
Certainly, is there a crate you already use that can handle these types of conversions, or should I introduce one to avoid the call to unsafe?